src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.cs

```diff
@@ -70,16 +87,46 @@ public static partial class JsonSerializer
                         }
                         else if (readStack.Current.IsProcessingDictionaryOrIDictionaryConstructible())
                         {
-                            HandleStartDictionary(options, ref readStack);
+                            readStack.EnqueueInitTask(InitTaskType.Dictionary);
                         }
                         else
                         {
-                            HandleStartObject(options, ref readStack);
+                            readStack.EnqueueInitTask(InitTaskType.Object);
                         }
                     }
                     else if (tokenType == JsonTokenType.EndObject)
                     {
-                        if (readStack.Current.Drain)
+                        if (readStack.Current.LastMetaProperty == MetadataPropertyName.Ref)
+                        {
+                            object refValue = readStack.GetReference(readStack.Current.MetadataId);
+                            // A reference object has its own stack frame.
+                            readStack.Pop();
+
+                            if (refValue == null)
+                            {
+                                HandleNull(ref reader, ref readStack);
+                            }
+                            else
+                            {
+                                // Reference to the entire Enumerable/Dictionary.
+                                if (!readStack.Current.CollectionPropertyInitialized && (readStack.Current.IsProcessingProperty(ClassType.Dictionary | ClassType.Enumerable)))
+                                    {
+                                    ApplyObjectToEnumerable(refValue, ref readStack, setPropertyDirectly: true);
+                                    readStack.Current.EndProperty();
+                                }
+                                else //Reference to an object within the enumerable/dictionary.
+                                {
+                                    ApplyObjectToEnumerable(refValue, ref readStack);
+                                }
+                            }
+                        }
+                        else if (readStack.Current.HandleWrappingBrace)
```

Comment:
I don't understand why we need `HandleWrappingBrace`

Reply:
When you have to add an $id to an array, the array gets transformed into an object:
```
{
    "$id": "1",
    "$values": []
}
```

So this flag is meant to set to true when this "Preserved Array" is passed in order to avoid handling the close curly brace as the end of an object.